pkg/ipmanager/ipam/controller.go

```diff
@@ -0,0 +1,236 @@
+			}
+			if annotations[submarinerIpamGlobalIp] == "" {
+				annotations[submarinerIpamGlobalIp] = "169.254.1.1"
+				service.SetAnnotations(annotations)
+				klog.V(4).Infof("Updating service %s", service.Name)
+				_, updateErr := i.kubeClientSet.CoreV1().Services(ns).Update(service)
+				return updateErr
+			}
+			return nil
+		})
+		if retryErr != nil {
+			i.serviceWorkqueue.Forget(obj)
+			return retryErr
+		}
+
+		i.serviceWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (i *Controller) processNextPod() bool {
+	obj, shutdown := i.podWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer i.podWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			i.podWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+		if i.excludeNamespaces[ns] {
+			i.podWorkqueue.Forget(obj)
+			return nil
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of Pod before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			pod, err := i.kubeClientSet.CoreV1().Pods(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				i.podWorkqueue.Forget(obj)
+				return fmt.Errorf("error retrieving submariner-ipam-controller pod object %s: %v", name, err)
+			}
+			annotations := pod.GetAnnotations()
+			if annotations == nil {
+				annotations = map[string]string{}
+			}
+			if annotations[submarinerIpamGlobalIp] == "" {
+				annotations[submarinerIpamGlobalIp] = "169.254.1.1"
+				pod.SetAnnotations(annotations)
+				klog.V(4).Infof("Updating pod %s", pod.Name)
+				_, updateErr := i.kubeClientSet.CoreV1().Pods(ns).Update(pod)
+				return updateErr
+			}
+			return nil
+		})
+		if retryErr != nil {
+			i.podWorkqueue.Forget(obj)
+			return retryErr
+		}
+
+		i.podWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (i *Controller) enqueueService(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	klog.V(6).Infof("Enqueueing service for ipam controller %v", obj)
+	i.serviceWorkqueue.AddRateLimited(key)
+}
+
+func (i *Controller) enqueuePod(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	klog.V(6).Infof("Enqueueing pod for ipam controller %v", obj)
+	i.podWorkqueue.AddRateLimited(key)
+}
+
+func (i *Controller) handleRemovedService(obj interface{}) {
+	klog.V(6).Infof("Handling remove service in ipamController %v", obj)
+	service := obj.(*k8sv1.Service)
+	klog.V(4).Infof("Removing service %v", service.Name)
```

Comment:
I hope in this method, we will mark the allocated ip (for Service and similarly in the next api for POD) as free. Please add a TODO comment.